File: src/main/java/com/arnav/attendanceapp/semester/service/SemesterService.java
package com.arnav.attendanceapp.semester.service;


import com.arnav.attendanceapp.calendar.service.CalendarDayService;
import com.arnav.attendanceapp.semester.entity.Semester;
import com.arnav.attendanceapp.semester.repo.SemesterRepo;
import org.springframework.stereotype.Service;

@Service
public class SemesterService {

    private final SemesterRepo semesterRepo;
    private final CalendarDayService  calendarDayService;

    public SemesterService(SemesterRepo semesterRepo,  CalendarDayService calendarDayService) {
        this.semesterRepo = semesterRepo;
        this.calendarDayService = calendarDayService;
    }

    public Semester createSemester(Semester semester) {
        Semester savedSemester = semesterRepo.save(semester);
        calendarDayService.generateSemesterCalendar(savedSemester);
        return savedSemester;
    }
}

File: backend/src/semester/semester.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Semester } from './semester.entity';
import { CreateSemesterDto } from './dto/create-semester.dto';
import { CalendarService } from '../calendar/calendar.service';

@Injectable()
export class SemesterService {
  constructor(
    @InjectRepository(Semester)
    private readonly semesterRepo: Repository<Semester>,
    private readonly calendarService: CalendarService,
  ) {}

  async create(dto: CreateSemesterDto): Promise<Semester> {
    const semester = this.semesterRepo.create(dto);
    const saved = await this.semesterRepo.save(semester);
    await this.calendarService.generateSemesterCalendar(saved);
    return saved;
  }

  findAll(): Promise<Semester[]> {
    return this.semesterRepo.find();
  }

  findOne(id: number): Promise<Semester> {
    return this.semesterRepo.findOneByOrFail({ id });
  }
}
